tls/s2n_psk.c

```diff
@@ -289,3 +289,45 @@ S2N_RESULT s2n_finish_psk_extension(struct s2n_connection *conn)
     GUARD_RESULT(s2n_psk_write_binder_list(conn, &partial_client_hello, client_hello));
     return S2N_RESULT_OK;
 }
+
+int s2n_connection_set_external_psks(struct s2n_connection *conn, struct s2n_pre_shared_key *psk_vec, size_t psk_vec_length)
+{
+    notnull_check(conn);
+    notnull_check(psk_vec);
+    gt_check(psk_vec_length, 0);
+
+    for (size_t i = 0; i < psk_vec_length; i++) {
+        /* Check for duplicate identities */
+        size_t array_len = conn->psk_params.psk_list.len;
+        for (size_t j = 0; j < array_len; j++) {
+            struct s2n_psk *psk = NULL;
+            GUARD_AS_POSIX(s2n_array_get(&conn->psk_params.psk_list, j, (void**) &psk));
+            if (psk->identity.size == psk_vec[i].identity.identity_length) {
+                S2N_ERROR_IF(s2n_constant_time_equals(psk->identity.data, psk_vec[i].identity.identity_data, psk->identity.size), S2N_ERR_DUPLICATE_IDENTITIES);
```

Comment:
There's no need to do a constant time comparison here. This is called by the server/client, not triggered by any remote action; the local PSKs are not a secret from the caller.

We also prefer ENSURE_POSIX to S2N_ERROR_IF. ENSURE_POSIX lets us assert a desired condition instead of asserting the opposite of the desired condition.